internal/storage/streamed_chunk_uploader.go

```diff
@@ -0,0 +1,206 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"cloud.google.com/go/storage"
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/storageutil"
+	"google.golang.org/api/googleapi"
+)
+
+type UploaderState int
+
+const (
+	// Default initial state in case of failure in initialization.
+	Uninitialized UploaderState = iota
+	// State of successful initialization, no writes so far.
+	Initialized
+	// No write is in progress, but there have been writes in the past.
+	Waiting
+	// Asynchronous write is in progress.
+	Writing
+	// Upload failed.
+	WriteError
+	// Has been finalized.
+	Destroyed
+)
+
+// A StreamedChunkUploader is an implementation of ChunkUploader
+// interface, which use storage.Writer from go-storage-client
+// for resumable upload.
+//
+// It also stores state of how many bytes have been sent to it
+// to write, how many have been successfully written by it so far,
+// and the current state of the uploader.
+type StreamedChunkUploader struct {
+	// Implements the following interfaces.
+	gcs.ChunkUploader
+
+	// Internal objects for business logic
+	wc         *storage.Writer
+	objectName string
+
+	// Attributes for providing updates to user
+	totalWriteInitiatedSoFar int64
+	totalWriteSucceededSoFar int64
+	userProgressFunc         func(int64)
+
+	// Internal state for lifecycle management
+	state UploaderState
+}
+
+// NewStreamedChunkUploader creates a new instace of StreamedChunkUploader
+// for the given inputs.
+func NewStreamedChunkUploader(ctx context.Context, obj *storage.ObjectHandle, req *gcs.CreateObjectRequest, writeChunkSize int, progressFunc func(int64)) (*StreamedChunkUploader, error) {
+	if obj == nil || req == nil {
+		return nil, fmt.Errorf("nil obj or req")
+	}
+
+	if obj.ObjectName() != req.Name {
+		return nil, fmt.Errorf("names of passed object-handle (%s) and CreateObjectRequest (%s) don't match", obj.ObjectName(), req.Name)
+	}
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		return nil, fmt.Errorf("request received for pre-existing object %s, supported only for new objects", req.Name)
+	}
+
+	if writeChunkSize <= 0 {
+		return nil, fmt.Errorf("writeChunkSize <= 0 not supported")
+	}
+
+	// raw initialization
+	sow := StreamedChunkUploader{}
+	sow.state = Uninitialized
+
+	// store references to necessary parameters
+	sow.objectName = obj.ObjectName()
+	sow.userProgressFunc = progressFunc
+
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// Chuck size for resumable upload is default i.e. 16MB.
+	wc := obj.NewWriter(ctx)
+	// NewWriter never returns nil, so no nil-check here.
```

Comment:
nit: add comments on top of line not below.

Reply:
The null check was needed later, so I added the comment later, to keep the flow for the reader. Have moved before the line now.